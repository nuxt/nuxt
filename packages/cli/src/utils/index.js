import path from 'path'
import exit from 'exit'

import { lock } from '@nuxt/utils'
import chalk from 'chalk'
import { isWindows } from 'std-env'
import { warningBox } from './formatting'

export const eventsMapping = {
  add: { icon: '+', color: 'green', action: 'Created' },
  change: { icon: isWindows ? '»' : '↻', color: 'blue', action: 'Updated' },
  unlink: { icon: '-', color: 'red', action: 'Removed' }
}

export function formatPath (filePath) {
  if (!filePath) {
    return
  }
  return filePath.replace(process.cwd() + path.sep, '')
}

/**
 * Normalize string argument in command
 *
 * @export
 * @param {String} argument
 * @param {*} defaultValue
 * @returns formatted argument
 */
export function normalizeArg (arg, defaultValue) {
  switch (arg) {
    case 'true': arg = true; break
    case '': arg = true; break
    case 'false': arg = false; break
    case undefined: arg = defaultValue; break
  }
  return arg
}

export function forceExit (cmdName, timeout) {
  if (timeout !== false) {
    const exitTimeout = setTimeout(() => {
      const msg = `The command 'nuxt ${cmdName}' finished but did not exit after ${timeout}s
This is most likely not caused by a bug in Nuxt
Make sure to cleanup all timers and listeners you or your plugins/modules start.
Nuxt will now force exit

${chalk.bold('DeprecationWarning: Starting with Nuxt version 3 this will be a fatal error')}`

      // TODO: Change this to a fatal error in v3
      process.stderr.write(warningBox(msg))
      exit(0)
    }, timeout * 1000)
    exitTimeout.unref()
  } else {
    exit(0)
  }
}

// An immediate export throws an error when mocking with jest
// TypeError: Cannot set property createLock of #<Object> which has only a getter
export function createLock (...args) {
  return lock(...args)
}
