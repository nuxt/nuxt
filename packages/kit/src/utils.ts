/** @since 3.9.0 */
export function toArray<T> (value: T | T[]): T[] {
  return Array.isArray(value) ? value : [value]
}

/**
 * Filter out items from an array in place. This function mutates the array.
 * `predicate` get through the array from the end to the start for performance.
 *
 * This function should be faster than `Array.prototype.filter` on large arrays.
 */
export function filterInPlace<T> (array: T[], predicate: (item: T, index: number, arr: T[]) => unknown) {
  for (let i = array.length; i--; i >= 0) {
    if (!predicate(array[i]!, i, array)) {
      array.splice(i, 1)
    }
  }
  return array
}

export const MODE_RE = /\.(server|client)(\.\w+)*$/
