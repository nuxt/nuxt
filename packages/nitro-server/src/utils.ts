import { fileURLToPath } from 'node:url'
import { dirname } from 'pathe'

export function toArray<T> (value: T | T[]): T[] {
  return Array.isArray(value) ? value : [value]
}

let _distDir = dirname(fileURLToPath(import.meta.url))
if (/(?:chunks|shared)$/.test(_distDir)) { _distDir = dirname(_distDir) }

export const distDir = _distDir
