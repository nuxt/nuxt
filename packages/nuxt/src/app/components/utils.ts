import { h } from 'vue'
import type { Component, RendererNode } from 'vue'
// eslint-disable-next-line
import { isString, isPromise, isArray, isObject } from '@vue/shared'
import destr from 'destr'

/**
 * Internal utility
 *
 * @private
 */
export const _wrapIf = (component: Component, props: any, slots: any) => {
  props = props === true ? {} : props
  return { default: () => props ? h(component, props, slots) : slots.default?.() }
}

// eslint-disable-next-line no-use-before-define
export type SSRBuffer = SSRBufferItem[] & { hasAsync?: boolean }
export type SSRBufferItem = string | SSRBuffer | Promise<SSRBuffer>

/**
 * create buffer retrieved from @vue/server-renderer
 *
 * @see https://github.com/vuejs/core/blob/9617dd4b2abc07a5dc40de6e5b759e851b4d0da1/packages/server-renderer/src/render.ts#L57
 * @private
 */
export function createBuffer () {
  let appendable = false
  const buffer: SSRBuffer = []
  return {
    getBuffer (): SSRBuffer {
      return buffer
    },
    push (item: SSRBufferItem) {
      const isStringItem = isString(item)
      if (appendable && isStringItem) {
        buffer[buffer.length - 1] += item as string
      } else {
        buffer.push(item)
      }
      appendable = isStringItem
      if (isPromise(item) || (isArray(item) && item.hasAsync)) {
        buffer.hasAsync = true
      }
    }
  }
}

const TRANSLATE_RE = /&(nbsp|amp|quot|lt|gt);/g
const NUMSTR_RE = /&#(\d+);/gi
export function decodeHtmlEntities (html: string) {
  const translateDict = {
    nbsp: ' ',
    amp: '&',
    quot: '"',
    lt: '<',
    gt: '>'
  } as const
  return html.replace(TRANSLATE_RE, function (_, entity: keyof typeof translateDict) {
    return translateDict[entity]
  }).replace(NUMSTR_RE, function (_, numStr: string) {
    const num = parseInt(numStr, 10)
    return String.fromCharCode(num)
  })
}

/**
 * helper for NuxtIsland to generate a correct array for scoped data
 */
export function vforToArray (source: any): any[] {
  if (isArray(source)) {
    return source
  } else if (isString(source)) {
    return source.split('')
  } else if (typeof source === 'number') {
    if (process.dev && !Number.isInteger(source)) {
      console.warn(`The v-for range expect an integer value but got ${source}.`)
    }
    const array = []
    for (let i = 0; i < source; i++) {
      array[i] = i
    }
    return array
  } else if (isObject(source)) {
    if (source[Symbol.iterator as any]) {
      return Array.from(source as Iterable<any>, item =>
        item
      )
    } else {
      const keys = Object.keys(source)
      const array = new Array(keys.length)
      for (let i = 0, l = keys.length; i < l; i++) {
        const key = keys[i]
        array[i] = source[key]
      }
      return array
    }
  }
  return []
}

export function getFragmentHTML (element: RendererNode | null) {
  if (element) {
    if (element.nodeName === '#comment' && element.nodeValue === '[') {
      return getFragmentChildren(element)
    }
    return [element.outerHTML]
  }
  return []
}

function getFragmentChildren (element: RendererNode | null, blocks: string[] = []) {
  if (element && element.nodeName) {
    if (isEndFragment(element)) {
      return blocks
    } else if (!isStartFragment(element)) {
      blocks.push(element.outerHTML)
    }

    getFragmentChildren(element.nextSibling, blocks)
  }
  return blocks
}

function isStartFragment (element: RendererNode) {
  return element.nodeName === '#comment' && element.nodeValue === '['
}

function isEndFragment (element: RendererNode) {
  return element.nodeName === '#comment' && element.nodeValue === ']'
}
const SLOT_PROPS_RE = /<div[^>]*nuxt-ssr-slot-name="([^"]*)" nuxt-ssr-slot-data="([^"]*)"[^/|>]*>/g

export function getSlotProps (html: string) {
  const slotsDivs = html.matchAll(SLOT_PROPS_RE)
  const data: Record<string, any> = {}
  for (const slot of slotsDivs) {
    const [_, slotName, json] = slot
    const slotData = destr(decodeHtmlEntities(json))
    data[slotName] = slotData
  }
  return data
}
