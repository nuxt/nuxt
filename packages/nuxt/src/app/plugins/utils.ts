const VALID_ISLAND_KEY_RE = /^[a-z][a-z\d-]*_[a-z\d]+$/i
/* @__PURE__ */
export function isValidIslandKey (key: string): boolean {
  return typeof key === 'string' && VALID_ISLAND_KEY_RE.test(key) && key.length <= 100
}
