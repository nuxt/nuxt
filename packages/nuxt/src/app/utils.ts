export function toArray<T> (value: T | T[] | undefined): T[] {
  return Array.isArray(value) ? value : (value ? [value] : [])
}
