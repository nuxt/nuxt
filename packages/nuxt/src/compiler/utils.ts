import { type ParsedStaticImport, findStaticImports, parseStaticImport } from 'mlly'
import { parseAndWalk, walk } from 'oxc-walker'
import type { ScanPlugin } from '@nuxt/schema'
import type { ParseResult } from 'oxc-parser'

/**
 * Creates a context object for scan plugins scoped to a specific file being scanned.
 * All plugins scanning this file will share the same context instance.
 */
export function createScanPluginContext (code: string, filePath: string) {
  // create the context for sharing the parse result between plugins
  let parseResult: ParseResult | null = null
  let parsedStaticImports: ParsedStaticImport[] | null = null

  const pluginScanThisContext: ThisParameterType<ScanPlugin['scan']> = {
    walkParsed: (...args) => {
      if (parseResult) {
        const options: Parameters<typeof walk>[1] = typeof args[0] === 'function' ? { enter: args[0] } : args[0]
        walk(parseResult.program, options)
        return parseResult
      }

      parseResult = parseAndWalk.call(null, code, filePath, args[0])
      return parseResult
    },
    getParsedStaticImports: () => {
      if (parsedStaticImports) {
        return parsedStaticImports
      }

      const imports = findStaticImports(code)
      parsedStaticImports = imports.map(i => parseStaticImport(i))
      return parsedStaticImports
    },
  }
  return pluginScanThisContext
}

export function matchWithStringOrRegex (value: string, matcher: string | RegExp) {
  if (typeof matcher === 'string') {
    return value === matcher
  } else if (matcher instanceof RegExp) {
    return matcher.test(value)
  }

  return false
}
