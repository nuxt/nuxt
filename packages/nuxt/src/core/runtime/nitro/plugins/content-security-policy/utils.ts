import { existsSync } from 'node:fs'
import { readFile, readdir } from 'node:fs/promises'
import { isAbsolute, join } from 'pathe'
import { createDefu } from 'defu'
import type { Nitro } from 'nitropack/types'
import type { ContentSecurityPolicyValue } from './types'

export function headerStringFromObject (optionValue: ContentSecurityPolicyValue | false) {
  // False value translates into empty header
  if (optionValue === false) {
    return ''
  }
  // Stringify the options passed as a JS object
  return Object.entries(optionValue)
    .filter(([, value]) => value !== false)
    .map(([directive, sources]) => {
      if (directive === 'upgrade-insecure-requests') {
        return 'upgrade-insecure-requests;'
      } else {
        const stringifiedSources = (typeof sources === 'string')
          ? sources
          : (sources as string[])
              .map(source => source.trim())
              .join(' ')
        return `${directive} ${stringifiedSources};`
      }
    })
    .join(' ')
}

export const defuReplaceArray = createDefu((obj, key, value) => {
  if (Array.isArray(obj[key]) || Array.isArray(value)) {
    obj[key] = value
    return true
  }
})

export function generateRandomNonce () {
  const array = new Uint8Array(18)
  crypto.getRandomValues(array)
  const nonce = btoa(String.fromCharCode(...array))
  return nonce
}

export async function generateHash (content: Buffer | string, hashAlgorithm: 'SHA-256' | 'SHA-384' | 'SHA-512') {
  let buffer: Uint8Array
  if (typeof content === 'string') {
    buffer = new TextEncoder().encode(content)
  } else {
    buffer = new Uint8Array(content)
  }
  const hashBuffer = await crypto.subtle.digest(hashAlgorithm, buffer)
  const base64 = btoa(String.fromCharCode(...new Uint8Array(hashBuffer)))
  const prefix = hashAlgorithm.replace('-', '').toLowerCase()
  return `${prefix}-${base64}`
}

export async function hashBundledAssets (nitro: Nitro) {
  const hashAlgorithm = 'SHA-384'
  const sriHashes: Record<string, string> = {}

  // Will be later necessary to construct url
  const { cdnURL: appCdnUrl = '', baseURL: appBaseUrl } = nitro.options.runtimeConfig.app

  // Go through all public assets folder by folder
  const publicAssets = nitro.options.publicAssets
  for (const publicAsset of publicAssets) {
    const { dir, baseURL = '' } = publicAsset

    if (existsSync(dir)) {
      // Node 16 compatibility maintained
      // Node 18.17+ supports recursive option on readdir
      // const entries = await readdir(dir, { withFileTypes: true, recursive: true })
      const entries = await readdir(dir, { withFileTypes: true })
      for (const entry of entries) {
        if (entry.isFile()) {
          // Node 16 compatibility maintained
          // Node 18.17+ supports entry.path on DirEnt
          // const fullPath = join(entry.path, entry.name)
          const path = join(dir, entry.name)
          const content = await readFile(path)
          const hash = await generateHash(content, hashAlgorithm)
          // construct the url as it will appear in the head template
          const fullPath = join(baseURL, entry.name)
          let url: string
          if (appCdnUrl) {
            // If the cdnURL option was set, the url will be in the form https://...
            const relativePath = isAbsolute(fullPath) ? fullPath.slice(1) : fullPath
            const abdsoluteCdnUrl = appCdnUrl.endsWith('/') ? appCdnUrl : appCdnUrl + '/'
            url = new URL(relativePath, abdsoluteCdnUrl).href
          } else {
            // If not, the url will be in a relative form: /_nuxt/...
            url = join('/', appBaseUrl, fullPath)
          }
          sriHashes[url] = hash
        }
      }
    }
  }

  return sriHashes
}
