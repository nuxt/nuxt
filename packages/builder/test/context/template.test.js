import hash from 'hash-sum'
import consola from 'consola'
import serialize from 'serialize-javascript'

import devalue from '@nuxt/devalue'
import { r, wp, wChunk, serializeFunction } from '@nuxt/utils'
import TemplateContext from '../../src/context/template'

jest.mock('lodash', () => ({
  ...jest.requireActual('lodash'),
  test: 'test lodash',
  warn: 'only once'
}))

describe('builder: buildContext', () => {
  const builder = {
    template: { files: ['template.js'] },
    globals: ['globals'],
    plugins: ['plugins'],
    relativeToBuild: jest.fn((...args) => `relativeBuild(${args.join(', ')})`)
  }
  const options = {
    features: { store: true },
    extensions: ['test', 'ext'],
    messages: { test: 'test message' },
    build: {
      splitChunks: { testSC: true }
    },
    dev: 'test_dev',
    test: 'test_test',
    debug: 'test_debug',
    vue: { config: 'test_config' },
    mode: 'test mode',
    router: { route: 'test' },
    env: 'test_env',
    head: 'test_head',
    store: 'test_store',
    globalName: 'test_global',
    css: ['test.css'],
    layouts: {
      'test-layout': 'test.template'
    },
    srcDir: 'test_src_dir',
    rootDir: 'test_root_dir',
    loading: 'test-loading',
    pageTransition: { name: 'test_trans' },
    layoutTransition: { name: 'test_layout_trans' },
    dir: ['test_dir'],
    ErrorPage: 'test_error_page'
  }

  test('should construct context', () => {
    const context = new TemplateContext(builder, options)
    expect(context).toMatchSnapshot()
  })

  test('should return object loading template options', () => {
    const context = new TemplateContext(builder, {
      ...options,
      loading: { name: 'test_loading' }
    })

    expect(context.templateVars.loading).toEqual({ name: 'test_loading' })
  })

  test('should return object loading template options', () => {
    const context = new TemplateContext(builder, options)
    const templateOptions = context.templateOptions
    expect(templateOptions).toEqual({
      imports: {
        serialize,
        serializeFunction,
        devalue,
        hash,
        r,
        wp,
        wChunk,
        _: {}
      },
      interpolate: /<%=([\s\S]+?)%>/g
    })
    expect(templateOptions.imports._.test).toEqual('test lodash')
    expect(templateOptions.imports._.warn).toEqual('only once')
    expect(consola.warn).toBeCalledTimes(1)
    expect(consola.warn).toBeCalledWith('Avoid using _ inside templates')
  })
})
