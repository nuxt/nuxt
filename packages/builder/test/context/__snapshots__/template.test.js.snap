// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`builder: buildContext should construct context 1`] = `
TemplateContext {
  "templateFiles": [
    "template.js",
  ],
  "templateVars": {
    "appPath": "./App.js",
    "buildIndicator": undefined,
    "components": {
      "ErrorPage": "relativeBuild(test_error_page)",
    },
    "css": [
      "test.css",
    ],
    "debug": "test_debug",
    "dir": [
      "test_dir",
    ],
    "env": "test_env",
    "extensions": "test|ext",
    "features": {
      "store": true,
    },
    "fetch": undefined,
    "globalName": "test_global",
    "globals": [
      "globals",
    ],
    "head": "test_head",
    "isDev": "test_dev",
    "isFullStatic": false,
    "isTest": "test_test",
    "layoutTransition": {
      "name": "test_layout_trans",
    },
    "layouts": {
      "test-layout": "test.template",
    },
    "loading": "relativeBuild(test_src_dir, test-loading)",
    "messages": {
      "test": "test message",
    },
    "mode": "test mode",
    "nuxtOptions": {
      "ErrorPage": "test_error_page",
      "build": {
        "splitChunks": {
          "testSC": true,
        },
      },
      "css": [
        "test.css",
      ],
      "debug": "test_debug",
      "dev": "test_dev",
      "dir": [
        "test_dir",
      ],
      "env": "test_env",
      "extensions": [
        "test",
        "ext",
      ],
      "features": {
        "store": true,
      },
      "globalName": "test_global",
      "head": "test_head",
      "layoutTransition": {
        "name": "test_layout_trans",
      },
      "layouts": {
        "test-layout": "test.template",
      },
      "loading": "test-loading",
      "messages": {
        "test": "test message",
      },
      "mode": "test mode",
      "pageTransition": {
        "name": "test_trans",
      },
      "rootDir": "test_root_dir",
      "router": {
        "route": "test",
      },
      "srcDir": "test_src_dir",
      "store": "test_store",
      "test": "test_test",
      "vue": {
        "config": "test_config",
      },
    },
    "pageTransition": {
      "name": "test_trans",
    },
    "plugins": [
      "plugins",
    ],
    "rootDir": "test_root_dir",
    "router": {
      "route": "test",
    },
    "splitChunks": {
      "testSC": true,
    },
    "srcDir": "test_src_dir",
    "store": "test_store",
    "uniqBy": [Function],
    "vue": {
      "config": "test_config",
    },
  },
}
`;
